### PR TITLE
State: Prevent logged-in user state from persisting for logged-out users

### PR DESCRIPTION
**Pull request by DavidRothstein**
This grew out of an issue discovered in https://github.com/Automattic/wp-calypso/pull/25626, but turned out to be a more fundamental bug.

Whenever you log out of Calypso, the entire state tree from the previous logged-in user gets written to the logged-out user's local storage.  This may have a variety of consequences.  One definite consequence is the bug report above ("Email me a login link" option missing on the login form).

### Steps to reproduce

1. Run Calypso locally using `ENABLE_FEATURES=no-force-sympathy npm start` (to mimic the production environment, in which the bug always occurs).
2. Log in to Calypso.
3. Log out.
4. Visit the login page again.

**Before the patch**: The "Email me a login link" (a.k.a. magic links) option will be missing from underneath the login form.  Furthermore, if you use your browser's development tools to look at `calypso_store` in the Indexed DB local storage, you'll see that the `redux-state-logged-out` entry has lots of data from the previous logged-in user (the `currentUser` and `sites` keys within that entry are particularly good places to look).

**After the patch**: The "Email me a login link" option should appear, and the  `redux-state-logged-out` entry in `calypso_store` should not contain any data from the previous logged-in user.

### Solution

As eventually determined in the discussion at https://github.com/Automattic/wp-calypso/pull/25626 (hat tip to several people, but especially @jsnajdr), this bug happens because although `user.clear()` (which is called on logout) clears the local storage, it doesn't do anything to change the Redux state in memory, or to prevent the `persistOnChange` method (particularly when called during the `beforeunload` event) from writing that now-stale Redux state to local storage later on.

There were a few possible ways to fix this discussed in the previous pull request.  Here, I settled on the solution of making the persistence code in `initial-state.js` detect cases where the user ID in the state doesn't match the user ID associated with the local storage key, and skip reading/writing the state when it does.  I chose this for a few reasons:

1. It seemed to be a more self-contained (and easy-to-write-tests-for) solution than having `user.clear()` trigger the persistence to be disabled.
2. It is more comprehensive, in the sense that if the current user ever changes via some other method besides `user.clear()`, the bug will still be prevented.
3. It doesn't just fix the bug when the state is being written, but also when it is being read. Although the latter isn't strictly necessary, it's extra protection, plus it also ensures that people experiencing the bug currently get it fixed right away, as opposed to waiting until the stored state expires - which can take 7 days, or potentially longer if they are interacting with the site during that time.

Some alternative fixes I didn't do:
- Stop persisting state for logged-out users altogether.  There were several people in the previous discussion who wanted to keep this feature in place.
- Actually make `user.clear()` remove the outdated Redux state from memory, so that it's automatically rebuilt for the new user without refreshing the page.  This would probably be the theoretically cleanest solution, but it led to way too many errors when I made a rudimentary attempt at it.  I did leave a code comment in place mentioning it though.

### CODE REVIEW

**Review by sirbrillig — COMMENTED** [review]
I appreciate the attention to detail in a fairly complex problem. As always, I'll defer to @dmsnell's review but this looks like a good fix to me. 

The only thing I'd like to see is testing instructions in the PR description.

**Comment by DavidRothstein** [comment]
> The only thing I'd like to see is testing instructions in the PR description.

Good point, I should have copied them over from the previous PR (and also expand them a bit, since the bug is deeper).  I'll do that now.

**Comment by dmsnell** [comment]
I'm sorry but I'm going to have to back out of review on this. I trust the others and I'm occupied with enough work that I can't gather the focus to give this the attention it deserves. Thanks for poking me on it!

**Comment by DavidRothstein** [comment]
I really do want to try to get a review from a Calypso "expert" to move forward here (since this addresses a pretty fundamental part of the system).  I added a few people as reviewers now who participated in the earlier discussion.  If any of you have time to review this, that would be great - thanks!

**Review by sirbrillig — APPROVED** [review]
Works as described. The more reviews the merrier, of course.

**Comment by blowery** [comment]
👋 Bit late to the party here. 

If the root of the problem is that we're hooking `beforeUnload` and writing out the redux store to the "wrong" key, could we just unhook the listener? We could add a dispatch to the logout method using the `redux-bridge` which the listener in initial-state could use to unhook the beforeUnload handler. 

I'm good with the checks in this PR, but the `user` module is going to go away eventually as we move everything to redux. I think unhooking the handler would be a bit more future proof. 

All that said, I'm happy to land this as is so we can get past the bug.

**Comment by blowery** [comment]
> Interesting... This file already does rely on the user module before this patch though (since it's used for determining the redux state key) so I'm guessing that rewrite will be at least somewhat necessary either way?

Yup, which is why it's fine for now. We'll have to redo it anyway.

>  I think what I'll do is leave this until early next week (it's a big enough change not to deploy right before a weekend anyway) and if there are no further reviews by then I'll deploy it.

Sounds like a plan!

**Comment by alisterscott** [comment]
I think this one is ready to merge

**Pull request merged**
